gh-85934: Use getattr_static when adding mock spec

melwitt opened the pull request:
When async magic method support was added to `unittest.mock.Mock` to address issue [#26467](https://bugs.python.org/issue26467), it introduced a `getattr` call [1] that causes class properties to be called when the class is used as a mock spec.

This caused a problem for a test in my project when running with Python 3.8 where previously the test worked OK with Python 3.6.

The test aims to verify that a class instance is not created if the called code path does not access the class property and thus the class will not create a heavy object unless it's needed (lazy create on access via `@property`).

As of Python 3.8, the `@property` is always called and is called by the mock spec process itself, even though the code path being tested does not access the class `@property`.

Here is a code snippet that illustrates the `@property` calling from the mock spec alone:

```python
class SomethingElse(object):
    def __init__(self):
        self._instance = None

    @property
    def instance(self):
        if not self._instance:
            self._instance = 'object'

...

    def test_property_not_called_with_spec_mock(self):
        obj = SomethingElse()
        self.assertIsNone(obj._instance)
        mock = Mock(spec=obj)
        self.assertIsNone(obj._instance)
```
```bash
$ ./python -m unittest -v unittest.test.testmock.testmock.MockTest.test_property_not_called_with_spec_mock
test_property_not_called_with_spec_mock (unittest.test.testmock.testmock.MockTest) ... FAIL

======================================================================
FAIL: test_property_not_called_with_spec_mock (unittest.test.testmock.testmock.MockTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vagrant/cpython/Lib/unittest/test/testmock/testmock.py", line 2173, in test_property_not_called_with_spec_mock
    self.assertIsNone(obj._instance)
AssertionError: 'object' is not None
```

In summary, since commit 77b3b7701a34ecf6316469e05b79bb91de2addfa, class properties are being called when adding mock specs and this can introduce side effects for tests that are verifying state affected by code inside a `@property`.

This replaces the `getattr` call with a `inspect.getattr_static` call to avoid executing class code as part of the mock speccing process.

[1] https://github.com/python/cpython/blob/fb2718720346c8c7a0ad2d7477f20e9a5524ea0c/Lib/unittest/mock.py#L492

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41768](https://bugs.python.org/issue41768) -->
https://bugs.python.org/issue41768
<!-- /issue-number -->


<!-- gh-issue-number: gh-85934 -->
* Issue: gh-85934
<!-- /gh-issue-number -->

melwitt committed:
bpo-41768: Use getattr_static when adding mock spec

Since commit 77b3b7701a34ecf6316469e05b79bb91de2addfa, class properties
are being called when adding mock specs and this can introduce side
effects for tests that are verifying state affected by code inside a
@property.

This replaces the getattr call with a inspect.getattr_static call to
avoid executing class code as part of the mock speccing process.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

terryjreedy commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -2156,6 +2166,12 @@ def trace(frame, event, arg):  # pragma: no cover
 
+    def test_property_not_called_with_spec_mock(self):
+        obj = SomethingElse()
+        self.assertIsNone(obj._instance)
```
```suggestion
        self.assertIsNone(obj._instance, msg='before mock')
```

terryjreedy commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -2156,6 +2166,12 @@ def trace(frame, event, arg):  # pragma: no cover
+        obj = SomethingElse()
+        self.assertIsNone(obj._instance)
+        mock = Mock(spec=obj)
+        self.assertIsNone(obj._instance)
```
```suggestion
        self.assertIsNone(obj._instance, msg='after mock')
```

terryjreedy commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -38,6 +38,16 @@ def cmeth(cls, a, b, c, d=None): pass
+    def instance(self):
+        if not self._instance:
+            self._instance = 'object'
+
```
```suggestion
        return self._instance
```
See below for added assert.

melwitt replied:
Hi, thank you for the review! I don't see an added assert that makes use of this part of the change, am I missing something?

terryjreedy replied:
The suggestion that I wrote but did not actually add until now.

melwitt replied:
A-ha, thanks! The new assert makes sense though I'm noticing I think this new line should be unindented one level. I don't see a way I can edit a suggestion when I add it to a batch.

melwitt replied:
Sigh, sorry, my mistake. I don't know what is wrong with my eyes today.

terryjreedy reviewed (dismissed):
I do not know the mock specification well enough to be sure of what it should do, except that modifying the underlying object is likely wrong, and that the mock should actually mock it.  So the code patch is likely correct and an additional test seems needed.

terryjreedy commented on Lib/unittest/test/testmock/testmock.py:
```diff
@@ -2156,6 +2166,12 @@ def trace(frame, event, arg):  # pragma: no cover
+        self.assertIsNone(obj._instance)
+        mock = Mock(spec=obj)
+        self.assertIsNone(obj._instance)
+
```
```suggestion
        self.assertEqual(mock.instance, 'instance')
```

melwitt committed:
Return object from SomethingElse.instance property

The goal of the property is provide access to the object and create it lazily on first access.

melwitt committed:
Apply suggestions from code review

Co-authored-by: Terry Jan Reedy <tjreedy@udel.edu>

melwitt committed:
Clean up whitespace damage and adjust assert

Fix accidental whitespace damage introduced from the web UI.

Slight change from the originally suggested assert as mocks do not
provide the implementation of the class, only the interface of it.

Instead, we can assert the return value from the actual object after
accessing the property.

terryjreedy reviewed (commented):
Changes match those requested.  I am hoping someone else reviews this, but I will read spec reference if not.

melwitt committed:
Merge branch 'main' of github.com:python/cpython into unittest.mock-getattr

Resolve merge conflicts with the current main branch.

melwitt committed:
Add newline at the end of NEWS entry

terryjreedy committed:
Merge branch 'main' into unittest.mock-getattr

carljm reviewed (changes requested):
If we do this, the change from `getattr` to `getattr_static` should be made in the existing loop on line 533; we shouldn't introduce a new copy of this loop.

melwitt committed:
Merge branch 'main' of https://github.com/python/cpython into unittest.mock-getattr

Resolve merge conflicts with the current main branch.

melwitt committed:
Move getattr_static call into existing loop

carljm reviewed (approved):
The two possible concerns here are that `getattr_static` is slower than `getattr`, and backward-compatibility. But I think mock speccing executing properties is a bug, and fixing that bug outweighs those two concerns. Also, @AlexWaygood made `getattr_static` faster in 3.12. And this behavior of mock spec executing properties was introduced relatively recently (with AsyncMock); that change was a back-compat break from previous longstanding behavior, and this is the fix for it.

Because this is a bug, I think the fix should also be backported.

arhadthedev commented on Misc/NEWS.d/next/Library/2020-09-16-16-53-06.bpo-41768.8_fWkC.rst:
```diff
@@ -0,0 +1 @@
+unittest.mock speccing no longer calls class properties
```
```suggestion
:mod:`unittest.mock` speccing no longer calls class properties.
Patch by Melanie Witt.
```

melwitt committed:
Update Misc/NEWS.d/next/Library/2020-09-16-16-53-06.bpo-41768.8_fWkC.rst

Co-authored-by: Oleg Iarygin <oleg@arhadthedev.net>

carljm reviewed (changes requested):
Currently async methods decorated with `@classmethod` or `@staticmethod` are detected as async attributes. This PR causes them not to be anymore. I think we need to preserve this behavior (and add tests for it.) We'll have to follow `__wrapped__` to the decorated method, and test it with `iscoroutinefunction`. We can do this using `inspect.unwrap`.

melwitt committed:
Handle async methods decorated with @classmethod or @staticmethod

melwitt committed:
Merge branch 'python:main' into unittest.mock-getattr
